File: leetcode/yahoo/TwoSum.cpp
#include<iostream>
#include<vector>
using namespace std;

vector<vector<int> > ans;
bool find(vector<int> a,int t,int l,int r)
{
    if(l>r) return false;
    int min=(l+r)/2;
    if(a[min]==t) 
    {
        return true;
    }
    if(a[min]>t) return find(a,t,l,min-1);
    if(a[min]<t) return find(a,t,min+1,r);
}
    
vector<vector<int> > twosum(vector<int> a,int target)
{
    sort(a.begin(),a.end());
    for(int i=0;i<a.size();i++)
    {
        int t=target-a[i];
        if(find(a,t,i+1,a.size()-1))
        {
            vector<int> temp;
            temp.push_back(a[i]);
            temp.push_back(t);
            ans.push_back(temp);
        }
    }
    return ans;
}

main()
{
    vector<int> a;
    for(int i=0;i<10;i++)
    {
        a.push_back(i);
    }
    vector<vector<int> > b=twosum(a,10);
    for(int i=0;i<b.size();i++)
    {
        for(int j=0;j<2;j++)
        {
            cout<<b[i][j]<<" ";
        }
        cout<<endl;
    }       
    system("pause");
}    

File: leetcode/leetcode/src/com/lintcode/twoSum.java
package com.lintcode;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * Created by shiyanghuang on 15/6/8.
 */
public class twoSum {
    /*
     * @param numbers : An array of Integer
     * @param target : target = numbers[index1] + numbers[index2]
     * @return : [index1 + 1, index2 + 1] (index1 < index2)
     */
    public int[] twoSum(int[] numbers, int target) {
        // write your code here
        ArrayList<Integer> origin = new ArrayList<Integer>(numbers.length);
        for (int i = 0; i < numbers.length; i++) {
            origin.add(numbers[i]);
        }
        Arrays.sort(numbers);
        int l = 0, r = numbers.length - 1;
        while (numbers[l] + numbers[r] != target) {
            if (numbers[l] + numbers[r] < target) {
                l++;
            } else {
                r--;
            }
        }
        int[] ans = new int[2];
        for (int i = 0; i < origin.size(); i++) {
            if (numbers[l] == origin.get(i)) {
                ans[0] = i + 1;
            }
            if (numbers[r] == origin.get(i)) {
                ans[1] = i + 1;
            }
        }
        if (ans[0] > ans[1]) {
            int tmp = ans[0];
            ans[0] = ans[1];
            ans[1] = tmp;
        }
        return ans;
    }

    public static void main(String[] args) {
        int[] a = {1,0,-1};
        int[] ans = new twoSum().twoSum(a,-1);
        System.out.print(ans[0] + " " + ans[1]);
    }
}
